### PR TITLE
feat: Remote agent and functions

### PR DESCRIPTION
**Pull request by richburdon**
Re: https://github.com/dxos/dxos/issues/3205

### Summary
📝🚀🌐

Added a design document for the remote functions feature in `docs/docs/design/remote-funcitons.md`. The document explains the spec, architecture, and use cases of the feature, which allows users and developers to run functions over ECHO spaces and external systems.

> _`Remote functions` -_
> _Designing for DXOS platform_
> _Winter of coding_

### Walkthrough
*  Add a new markdown document that describes the design of remote functions ([link](https://github.com/dxos/dxos/pull/3217/files?diff=unified&w=0#diff-6136e1e653a2f7b289920a22e4c6b42842cf3e37180fb00638e7b519138a11f9R1-R65))

### CODE REVIEW

File: docs/docs/design/remote-funcitons.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Remote Functions
+
+Remote functions are part of the DXOS extenisbility framework that enables developers to deploy long-running agents and invoke short-lived functions that have access to Spaces.
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:3]: i would avoid calling everything extensibility to help focus the words around just the frontend sdk. We want to be able to distinguish between user-space (apps, functions) and kernel-space (extensibility) domains, and would likely want to hold back on infrastructure extensibility while delivering user-space value early (i.e. this box runs your functions if you declare them like this. here's the tool to run it with ... .)

Comment (zhenyasav) [reply docs/docs/design/remote-funcitons.md:3]: which is distinct from "you can also extend this box to handle other scenarios we don't support" which is likely separable for a first hypothesis

File: docs/docs/design/remote-funcitons.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+## Spec
+
+- Run long-lived functions that operate over users' ECHO spaces and interace with external systems.
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:7]:
```suggestion
- Run long-lived functions that operate over users' ECHO spaces and interface with external systems.
```

File: docs/docs/design/remote-funcitons.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# Remote Functions
+
+Remote functions are part of the DXOS extenisbility framework that enables developers to deploy long-running agents and invoke short-lived functions that have access to Spaces.
+
+## Spec
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:5]:
```suggestion
## Scenarios
```

File: docs/docs/design/remote-funcitons.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+- Run long-lived functions that operate over users' ECHO spaces and interace with external systems.
+- Isolation of individual functions.
+- Controlled access to resources (e.g., storage, network). Enables "pure" stateless funcitons.
+- Functions triggered by external events (e.g., signaling, HALO/ECHO mutations, timers, other functions).
+- Global registration and discovery (DMG).
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:11]: use a scenario to make it clear why this is a thing?
e.g.: users can delegate invocation to functions they don't control for security and compliance purposes (auth flows, signing flows, on-behalf-of authorizations)? (I don't think this is a scenario we're interested in here though)

File: docs/docs/design/remote-funcitons.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+- Life cycle management (i.e., install, start, pause, stop, destory).
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:12]:
```suggestion
- Users have visibility and control over lifecycle of functions running on their node (install, start, stop, destroy, etc)
```

File: docs/docs/design/remote-funcitons.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+- Universal search across user's (or groups)'s spaces.
+- Local training of LLM over a set of spaces.
+- Bridges to external data sources (e.g., Protonmail, Calendar, enterprise SQL database, web crawler, OAuth APIs.)
+
+
+## Goals
+
+- Reuse of existing open source systems.
+- Memory efficient runtime enabling high density.
+- Support multiple runtimes (e.g., WASI; potential to run within browser).
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:27]:
```suggestion
- Portability, compatibility across multiple platforms and "execution anywhere".
```

File: docs/docs/design/remote-funcitons.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+- Local testing environment.
+- Ability to self-host (e.g., not require require complex K8s infrastructure)
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:29]:
```suggestion
- Configuration-free, flexible self-hosting (e.g., not require require learning about complex K8s infrastructure or concepts like docker compose etc)
```

File: docs/docs/design/remote-funcitons.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+
+## Non-Goals
+
+- Coordination of funtions (or function state) runtime across multiple VMs (i.e., homogenous network).
+- Scale-on-demand: Functions run within the context of a deployed agent; it is not expected to support running 100s of function invocations. "Heavy lifting" might be achieved via external systems, which are out of bound.
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:35]:
```suggestion
- High scale. This system is viable within <1000rps and is not expected to handle high volume or high availability scenarios on it's own. 
```

File: docs/docs/design/remote-funcitons.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+- **Agent**: 
+  - DXOS peer (`@dxos/client`) with HALO (identity) and access to Spaces.
+  - Context for function invocation.
+- **Function**:
+  - Short-lived function ("lambda").
+  - Invoked with context (Agent, resource APIs) and event.
+  - DMG record (module) defines a group of Functions, and their runtime requirements (runtime, resources, versioned APIs, and dependency graph -- which may include other functions.) Module definition includes the CID or the resource bundle.
+  - Resources may include "system" services (e.g., storage, network access) and external endpoints (e.g., Local network services like IMAP server, LLM, or remote services like OAuth gateways.)
+- **DMG**:
+  - Federated registry that enables publishing and discover of funcitons.
```
Comment (zhenyasav) [inline docs/docs/design/remote-funcitons.md:69]:
```suggestion
  - Federated registry that enables publishing and discovery of funcitons.
```

**Comment by zhenyasav** [comment]
fyi forever https://www.npmjs.com/package/forever

File: packages/sdk/client-protocol/src/config.ts [hunk 10]
```diff
@@ -14,9 +14,25 @@ export const DEFAULT_SHELL_CHANNEL = 'dxos:shell';
...
+
+const HOME = typeof process !== 'undefined' ? process?.env?.HOME : '';
+
+// TODO(burdon): Generalize since currently NodeJS only.
+export const DX_CONFIG = `${HOME}/.config/dx`;
+export const DX_CACHE = `${HOME}/.cache/dx`;
+export const DX_DATA = `${HOME}/.local/share/dx`;
+
+// TODO(burdon): Override via XDG_RUNTIME_DIR
+export const DX_RUNTIME = '/tmp/dx/run';
```
Comment (dmaretskyi) [inline packages/sdk/client-protocol/src/config.ts:30]: windows)

**Pull request merged**